modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/TargetRequest.java

```diff
@@ -139,108 +137,135 @@ public void start(NHttpClientConnection conn) throws IOException, HttpException
-            boolean forceContentLengthCopy = requestMsgCtx.isPropertyTrue(
-                                                                   PassThroughConstants.COPY_CONTENT_LENGTH_FROM_INCOMING);
-                                
-            if (forceContentLength) {
-                entity.setChunked(false);
-                if (forceContentLengthCopy && contentLength > 0) {
-                    entity.setContentLength(contentLength);
-                } 
-            }else{
-             if (contentLength != -1) {
+        boolean forceContentLength = 
+                requestMsgCtx.isPropertyTrue(NhttpConstants.FORCE_HTTP_CONTENT_LENGTH);
+        boolean forceContentLengthCopy = 
+                requestMsgCtx.isPropertyTrue(PassThroughConstants.COPY_CONTENT_LENGTH_FROM_INCOMING);
+                            
+        if (forceContentLength) {
+            entity.setChunked(false);
+            if (forceContentLengthCopy && contentLength != -1) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Set ContentLength : " + contentLength);
+                }
+                entity.setContentLength(contentLength);
+            } else if (!hasEntityBody) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Set BasicHttpRequest Object");
+                }
+                request = new BasicHttpRequest(method, path, version != null ? version : HttpVersion.HTTP_1_1);
+            }
+        } else {
+            if (contentLength != -1) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Set ContentLength : " + contentLength);
+                }
                 entity.setChunked(false);
                 entity.setContentLength(contentLength);
             } else {
-                entity.setChunked(chunk);
-            }
+                if (hasEntityBody) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Set chunked : " + chunk);
+                    }
+                    entity.setChunked(chunk);
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Set BasicHttpRequest Object");
+                    }
+                    request = new BasicHttpRequest(method, path, version != null ? version : HttpVersion.HTTP_1_1);
+                }
             }
-            
-            
-            ((BasicHttpEntityEnclosingRequest) request).setEntity(entity);
-           
-        } else {
-            request = new BasicHttpRequest(method, path,
-                    version != null ? version : HttpVersion.HTTP_1_1);
         }
 
         Set<Map.Entry<String, LinkedHashSet<String>>> entries = headers.entrySet();
         for (Map.Entry<String, LinkedHashSet<String>> entry : entries) {
-             if (entry.getKey() != null) {
+            if (entry.getKey() != null) {
                 Iterator<String> i = entry.getValue().iterator();
```

Comment:
Shall we use some meaningful name for this variable instead of i, Even this is there from the beginning?

Reply:
fixed